### PR TITLE
 Implements a % cap on staking rewards from era inflation

### PR DESCRIPTION
**Pull request by gpestana**
This PR implements an (optional) cap of the era inflation that is allocated to staking rewards. The remaining is minted directly into the [`RewardRemainder`](https://github.com/paritytech/polkadot-sdk/blob/fb0fd3e62445eb2dee2b2456a0c8574d1ecdcc73/substrate/frame/staking/src/pallet/mod.rs#L160) account, which is the treasury pot account in Polkadot and Kusama.

The staking pallet now has a percent storage item, `MaxStakersRewards`, which defines the max percentage of the era inflation that should be allocated to staking rewards. The remaining era inflation (i.e. `remaining = max_era_payout - staking_payout.min(staking_payout * MaxStakersRewards))` is minted directly into the treasury. 

The `MaxStakersRewards` can be set by a privileged origin through the `set_staking_configs` extrinsic.

**To finish**
- [x] run benchmarks for westend-runtime

Replaces https://github.com/paritytech/polkadot-sdk/pull/1483
Closes https://github.com/paritytech/polkadot-sdk/issues/403

### CODE REVIEW

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3757912

**Comment by gpestana** [comment]
bot bench polkadot-pallet --pallet=pallet_staking --runtime polkadot

**Comment by gpestana** [comment]
bot bench substrate-pallet --pallet=pallet_staking

File: polkadot/runtime/common/src/impls.rs
Comment (liamaharon) [reply polkadot/runtime/common/src/impls.rs:95]: also some comments explaining what this math is doing could be helpful for newbies reading it 🐥

File: substrate/frame/staking/src/lib.rs
Comment (liamaharon) [inline substrate/frame/staking/src/lib.rs:895]: safe math?

File: polkadot/runtime/common/src/impls.rs
Comment (Ank4n) [inline polkadot/runtime/common/src/impls.rs:67]: Nit: It is a little bit weird to have min_remainder especially considering the docs for era_payout. Personally I would suggest to instead use `max_fraction_stakers`.
```
/// Determine the payout for this era.
///
/// Returns the amount to be paid to stakers in this era, as well as whatever else should be
/// paid out ("the rest").
```

File: substrate/frame/staking/src/mock.rs
Comment (Ank4n) [reply substrate/frame/staking/src/mock.rs:216]: Or may be `Zero::zero()`.

File: substrate/frame/staking/src/pallet/mod.rs
Comment (Ank4n) [inline substrate/frame/staking/src/pallet/mod.rs:492]: As above, I feel more convinced that for a staking pallet, `MaxStakedRewards` would make more sense.

**Review by Ank4n — APPROVED** [review]
A small nit but otherwise looks good.

File: polkadot/runtime/common/src/impls.rs
Comment (juangirini) [reply polkadot/runtime/common/src/impls.rs:95]: interestingly enough, there was some unsafe math already there

File: polkadot/runtime/kusama/src/weights/pallet_staking.rs
Comment (juangirini) [inline polkadot/runtime/kusama/src/weights/pallet_staking.rs:800]: leaving this comment here to not forget about it at the end

File: polkadot/runtime/polkadot/src/weights/pallet_staking.rs
Comment (juangirini) [inline polkadot/runtime/polkadot/src/weights/pallet_staking.rs:800]: ☝️ idem

File: polkadot/runtime/westend/src/weights/pallet_staking.rs
Comment (juangirini) [inline polkadot/runtime/westend/src/weights/pallet_staking.rs:800]: ☝️ idem

File: polkadot/runtime/common/src/impls.rs
Comment (gpestana) [reply polkadot/runtime/common/src/impls.rs]: Yes, I think this is a good idea. It also simplifies the era payout math. I have changed the PR accordingly.

**Comment by gpestana** [comment]
bot bench substrate-pallet --pallet=pallet_staking

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3779340

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3779482

File: polkadot/runtime/common/src/impls.rs
Comment (gpestana) [reply polkadot/runtime/common/src/impls.rs]: Safe math used now and added a few comments.

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3779616

**Comment by gpestana** [comment]
bot bench polkadot-pallet --pallet=pallet_staking --runtime polkadot

File: substrate/frame/staking/src/lib.rs
Comment (kianenigma) [inline substrate/frame/staking/src/lib.rs:854]: I invite you to document this trait to say what each parameter is doing. We might then conclude that we are polluting a _trait's API_ based on _one_ of the implementations. 

Nonetheless, given that we have agreed to rewrite these pallets it is fine.

Comment (kianenigma) [reply substrate/frame/staking/src/lib.rs:854]: In fact, you wouldn't have needed this at all if you had used `parameter_types! {storgae}` to store the percentage, right? https://github.com/paritytech/polkadot-sdk/pull/1660/files#r1335491918

**Review by kianenigma — APPROVED** [review]
There's a couple of API design nitpicks that I don't want them to block this PR, but I think it would be good to discuss them and make sure we are on the same page about them. Given that this pallet is already passed the point of remaining generic, it is fine, but a typical pallet that is meant to be reusable should not exhibit the type of design that is used. 

Needs auditing as well? 

Also, most of these changes need to go into the fellowship repo? see https://github.com/polkadot-fellows/runtimes/issues/27.

**Review by kianenigma — CHANGES_REQUESTED** [review]
Ahh, sorry, one main fix that need to happen before any merge is to make sure this is backwards compatible. If merged as-is, I think it will redirect all rewards to treasury. It should be in fact the opposite: This should be a no-op upon enactment, until governance changes the parameter. The default should be 100% for that.

**Comment by gpestana** [comment]
> Ahh, sorry, one main fix that need to happen before any merge is to make sure this is backwards compatible. If merged as-is, I think it will redirect all rewards to treasury. It should be in fact the opposite: This should be a no-op upon enactment, until governance changes the parameter. The default should be 100% for that.

That's a good catch, thanks! The PR was initially using the `MinRemainingPayout` which was default to 0. This should be changed in the new iteration which sets the max validator payout (which by default is 100, correct).

File: substrate/frame/staking/src/pallet/mod.rs
Comment (gpestana) [reply substrate/frame/staking/src/pallet/mod.rs:492]: One of the goals of this change is to expose is to expose the minimum % of the era inflation that goes directly into the treasury, thus the storage item instead of the runtime parameters.

File: substrate/frame/staking/src/lib.rs
Comment (gpestana) [reply substrate/frame/staking/src/lib.rs:854]: > In fact, you wouldn't have needed this at all if you had used `parameter_types! {storgae}` to store the percentage, right? https://github.com/paritytech/polkadot-sdk/pull/1660/files#r1335491918

That's correct. One of the main goals though is to expose the min % of rewards going directly to the treasury through governance, and that value needs to be propagated from the staking pallet to the `era_payout` impl. So I think that the trait needs to be changed, unless I'm missing something?

We could also keep the current `EraPayout` trait and formula as-is and then adjust the payout in the caller site (i.e. the staking pallet), based on the `MaxStakersPayout`. That's basically what we had in the 1st version of the PR but it also doesn't feel super clean.

**Comment by gpestana** [comment]
> Ahh, sorry, one main fix that need to happen before any merge is to make sure this is backwards compatible.

@kianenigma good catch, this has been addressed in the latest commit. Now the default max staking rewards is 100% (i.e. if `MaxStakedRewards` storage item is `None`).

As discussed elsewhere and for the sake of visibility, I agree that exposing the `max_validators_rewards` to `EraPayout` is crossing the abstraction boundaries and the era payout should only calculate what is the `max_era_payout` and then let the caller decide how to distribute the max inflation, calculate the remainder, etc however, I think it's sensible to postpone these changes when we implement a way to [split inflation to more origins than staking](https://hackmd.io/DIRWakzwQiirf4rGb8PO3Q?view) and/or as we prepare the move towards staking parachain.

For now, I think that it's OK to change the `EraPayout` trait to consider the max rewards payout (based on a storage item in the staking pallet) so we can encapsulate the max era staking payout and remainder in the implementor of `EraPayout`.

There is another sensible option for this PR and the short term which consists of not touching `EraPayout` implementation and then adjust the returned `(staking_payout, remainder)` in the caller site (staking pallet) based on the max % of staking rewards.

In any case, we will change this flow considerably in the future so I think the current PR is OK and I can open an issue with the notes from this discussion for when we improve the current impl.

**Comment by gpestana** [comment]
@kianenigma after playing with the storage parameter type, I agree with you that this can be implemented in a much cleaner/abstracted way. Closing this PR in lieu of https://github.com/paritytech/polkadot-sdk/pull/1859

**Comment by gpestana** [comment]
Reopening since https://github.com/paritytech/polkadot-sdk/pull/1859 relies on `root_track` to update the max validator rewards.

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3962696

File: substrate/frame/staking/src/lib.rs
Comment (gpestana) [reply substrate/frame/staking/src/lib.rs]: I changed the PR so that the `EraPayout` trait does not need to change. The adjustments to the inflation destination (i.e. validators rewards or remainder aka treasury) are now performed at the staking pallet level, not the era_payout implementation.

This way we don't pollute the EraPayout trait with staking-related logic.

File: substrate/frame/staking/src/pallet/mod.rs
Comment (kianenigma) [inline substrate/frame/staking/src/pallet/mod.rs:1792]: Could also be part of `set_staking_configs`.

**Review by kianenigma — APPROVED** [review]
Please make an issue that once we have a better parameters pallet, such that we can parameterize this in the runtime, this should be moved to a runtime parameter, and directly applied to whoever is sitting on the other side of `T::EraPayout`

**Comment by gpestana** [comment]
Closing due to https://github.com/paritytech/polkadot-sdk/issues/403#issuecomment-1794635026

**Comment by gpestana** [comment]
bot fmt

**Comment by gpestana** [comment]
bot bench polkadot-pallet --pallet=pallet_staking --runtime polkadot

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4857443

**Comment by gpestana** [comment]
bot bench substrate-pallet --pallet=pallet_staking

**Comment by gpestana** [comment]
bot bench substrate-pallet --pallet=pallet_staking

**Comment by gpestana** [comment]
bot bench polkadot-pallet --pallet=pallet_staking --runtime westend

File: substrate/frame/staking/src/pallet/mod.rs
Comment (kianenigma) [reply substrate/frame/staking/src/pallet/mod.rs:1983]: conclusion on this?

**Review by kianenigma — COMMENTED** [review]
Still LGTM.

**Comment by gpestana** [comment]
> Still LGTM.

@kianenigma, we are still waiting for the audit on this PR, it seems that @0xjaypi has fast tracked this audit and the new estimate is around mid next week.

File: substrate/frame/staking/src/pallet/mod.rs
Comment (gpestana) [reply substrate/frame/staking/src/pallet/mod.rs]: Removed this callable and now the max staked rewards can be set via the `set_staking_configs`. To keep track of removing this (and others) configs once we have the params pallet: https://github.com/paritytech/polkadot-sdk/issues/2874

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5050340

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5050482

**Comment by gpestana** [comment]
bot fmt

**Comment by gpestana** [comment]
bot bench polkadot-pallet --pallet=pallet_staking --runtime westend

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5050626

**Comment by gpestana** [comment]
bot cancel 2-50f51a7a-9fe2-4098-a268-af52c5e61823

**Comment by gpestana** [comment]
bot bench polkadot-pallet --pallet=pallet_staking --runtime westend

**Comment by Polkadot-Forum** [comment]
This pull request has been mentioned on **Polkadot Forum**. There might be relevant details there:

https://forum.polkadot.network/t/polkadots-economics-tools-to-shape-the-forseeable-future/8708/1

**Comment by Polkadot-Forum** [comment]
This pull request has been mentioned on **Polkadot Forum**. There might be relevant details there:

https://forum.polkadot.network/t/proposal-for-adjusting-polkadots-inflation-system-reducing-issuance-and-complexity/9157/1

**Pull request merged**
